File: src/main/java/komu/blunt/core/CoreLetExpression.java
package komu.blunt.core;

import komu.blunt.analyzer.VariableReference;
import komu.blunt.asm.Assembler;
import komu.blunt.asm.Instructions;
import komu.blunt.asm.Linkage;
import komu.blunt.asm.Register;

import static com.google.common.base.Preconditions.checkNotNull;

public final class CoreLetExpression extends CoreExpression {

    private final VariableReference var;
    private final CoreExpression value;
    private final CoreExpression body;

    public CoreLetExpression(VariableReference var, CoreExpression value, CoreExpression body) {
        this.var = checkNotNull(var);
        this.value = checkNotNull(value);
        this.body = checkNotNull(body);
    }

    @Override
    public Instructions assemble(Assembler asm, Register target, Linkage linkage) {
        Instructions instructions = new Instructions();
        instructions.append(new CoreSetExpression(var, value).assemble(asm, target, Linkage.NEXT));
        instructions.append(body.assemble(asm, target, linkage));
        return instructions;
    }
}

File: src/main/kotlin/komu/blunt/core/CoreLetExpression.kt
package komu.blunt.core

import komu.blunt.analyzer.VariableReference
import komu.blunt.asm.*

class CoreLetExpression(private val variable: VariableReference,
                        private val value: CoreExpression,
                        private val body: CoreExpression) : CoreExpression() {

    override fun assemble(asm: Assembler, target: Register, linkage: Linkage) =
        instructions {
            instructionsOf(CoreSetExpression(variable, value).assemble(asm, target, Linkage.NEXT))
            instructionsOf(body.assemble(asm, target, linkage))
        }

    override fun simplify() =
        // TODO: if value is constant, propagate it into body
        CoreLetExpression(variable, value.simplify(), body.simplify())

    override fun toString() = "(let ($variable $value) $body)"
}

